core/trino-main/src/main/java/io/trino/operator/PageValidations.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator;
+
+import com.google.common.base.Joiner;
+import io.trino.spi.Page;
+import io.trino.spi.TrinoException;
+import io.trino.spi.block.Block;
+import io.trino.spi.block.ValueBlock;
+import io.trino.spi.type.Type;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Supplier;
+
+import static io.trino.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+
+public final class PageValidations
+{
+    private PageValidations() {}
+
+    public static void validateOutputPageTypes(Page page, List<Type> expectedTypes, Supplier<String> debugContextSupplier)
+    {
+        if (page.getPositionCount() == 0) {
+            // Allow empty pages without blocks. It is valid for connectors/table functions to produce empty pages with empty block list 1Code has comments. Press enter to view.
+            // interleaved with pages with data. Example of such behavior can be found in TableChangesFunctionProcessor (this is not exactly 1Code has comments. Press enter to view.
+            // related to SourcePage, but illustrates the contract, which is the same for Pages and SourcePages). 1Code has comments. Press enter to view.
+            // Empty pages are filtered out by Driver.
+            return;
+        }
+
+        if (page.getChannelCount() != expectedTypes.size()) {
+            throw new TrinoException(
+                    GENERIC_INTERNAL_ERROR,
+                    "Invalid number of channels; got %s expected %s; context=%s; blocks=%s; types=%s".formatted(
+                            page.getChannelCount(),
+                            expectedTypes.size(),
+                            debugContextSupplier.get(),
+                            blocksDebugInfo(page),
+                            expectedTypes));
+        }
+
+        List<String> mismatches = null;
```

Comment:
`// Avoid allocation on the happy path`

(btw `new ArrayList(0)` could perhaps also avoid allocation if the list object itself is on the stack... no change requested)